packages/iov-core/src/servercore.ts

```diff
@@ -0,0 +1,80 @@
+import { ChainId, PublicIdentity, TransactionId, UnsignedTransaction } from "@iov/bcp-types";
+import { UserProfile, WalletId } from "@iov/keycontrol";
+
+import { MultiChainSigner } from "./multichainsigner";
+
+function publicIdentityEquals(left: PublicIdentity, right: PublicIdentity): boolean {
+  return (
+    left.chainId === right.chainId &&
+    left.pubkey.algo === right.pubkey.algo &&
+    left.pubkey.data.length === right.pubkey.data.length &&
+    left.pubkey.data.every((value, index) => value === right.pubkey.data[index])
+  );
+}
+
+export class ServerCore {
+  private readonly signer: MultiChainSigner;
+  private readonly profile: UserProfile;
+
+  constructor(profile: UserProfile, signer: MultiChainSigner) {
+    this.signer = signer;
+    this.profile = profile;
+  }
+
+  public async getIdentities(
+    _: string,
+    chainIds: ReadonlyArray<ChainId>,
+  ): Promise<ReadonlyArray<PublicIdentity>> {
+    const matchingIdentities = this.allIdentities().filter(identity => {
+      return chainIds.some(chainId => identity.chainId === chainId);
+    });
+
+    // TODO: ask user for permission and allow selection of identities
```

Comment:
Maybe we pass in a "authorize" callback in the constructor, like

```
constructor(
  profile: UserProfile, 
  authorizeView: (reason: string, chains: ReadonlyArray<ChainId>) => Promise<boolean>,
  authorizeSign: (reason: string, transaction: UnsignedTransaction) => Promise<boolean>)
```

Or we could wrap ServerCore with another class that exposes the same public methods, yet performs some custom UI checks before calling this class?

Just ideas....

Reply:
Good idea, I prefer the callbacks. Added the approach to https://github.com/iov-one/iov-core/issues/651